src/query/disjunction.rs

```diff
@@ -0,0 +1,284 @@
+use std::cmp::{Ordering, Reverse};
+use std::collections::BinaryHeap;
+
+use crate::query::score_combiner::DoNothingCombiner;
+use crate::query::{ScoreCombiner, Scorer};
+use crate::{DocId, DocSet, Score, TERMINATED};
+
+pub struct DisjunctionScorer<TScorer, TScoreCombiner = DoNothingCombiner> {
+    chains: MinHeap<TScorer>,
+    minimum_matches_required: usize,
+    score_combiner: TScoreCombiner,
+
+    doc: DocId,
+    score: Score,
+    is_end: bool,
+}
+
+// TODO: Try reconstruct, it's ugly.
+type MinHeap<T> = BinaryHeap<Reverse<ScorerWrapper<T>>>;
+
+#[repr(transparent)]
+struct ScorerWrapper<T>(T);
```

Comment:
I suspect we would be better off caching the doc. 
T is sometimes a Box, which implies dynamic dispatch on every call of `.doc()`

```suggestion
struct ScorerWrapper<T>(T);
```